### PR TITLE
[WIP] Sklearndev

### PR DESCRIPTION
**Pull request by DrAnaximandre**
Integration of scikit-learn backend

---

---

This change is [Reviewable](https://reviewable.io/reviews/tboquet/python-alp/8)

### CODE REVIEW

File: src/alp/backend/sklearn_backend.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    for d, dv in zip(data, data_val):
+        model.fit(d['X'], d['y'], *args, **kwargs)
+        predondata.append(model.predict(data[0]))
+        predonval.append(model.predict(dv[0]))
+
+    # Validates the model
+    # So far, only the mae is supported.
+    from sklearn.metrics import mean_absolute_error
+    metrics.append(mean_absolute_error)
+    for metric in metrics:
```
Comment (tboquet) [inline src/alp/backend/sklearn_backend.py]: Codacy Issue found: [Unused variable 'metric'](https://www.codacy.com/app/thomas-boquet/python-alp/file/2670067469/issues/source?bid=3232749&fileBranchId=3332364#l253)

**Comment by tboquet** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/alp/backend/sklearn_backend.py  1
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/thomas-boquet/python-alp/pullRequest?prid=248711&bid=3332364)

File: src/alp/backend/sklearn_backend.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
```
Comment (tboquet) [inline src/alp/backend/sklearn_backend.py]: Codacy Issue found: [Unused variable 'metric'](https://www.codacy.com/app/thomas-boquet/python-alp/file/2670180122/issues/source?bid=3232749&fileBranchId=3332364#l253)

**Comment by tboquet** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/alp/backend/sklearn_backend.py  1
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/thomas-boquet/python-alp/pullRequest?prid=248711&bid=3332364)

File: src/alp/backend/sklearn_backend.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    for d, dv in zip(data, data_val):
+        model.fit(d['X'], d['y'], *args, **kwargs)
+        predondata.append(model.predict(d['X']))
+        predonval.append(model.predict(dv['X']))
+
+    # Validates the model
+    # So far, only the mae is supported.
+    from sklearn.metrics import mean_absolute_error
+    metrics.append(mean_absolute_error)
+    for metric in metrics:
```
Comment (tboquet) [inline src/alp/backend/sklearn_backend.py]: Codacy Issue found: [Unused variable 'metric'](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686525109/issues/source?bid=3232749&fileBranchId=3332364#l256)

File: tests/backend/test_sklearn_backend.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+    data["y"] = y_train
+
+    data_val["X"] = X_test
+    data_val["y"] = y_test
+
+    lr = LinearRegression(normalize=False)
+
+    model_dict = dict()
+    model_dict['model_arch'] = SKB.to_dict_w_opt(lr)
+    model_deserialized = SKB.model_from_dict_w_opt(model_dict['model_arch'])
```
Comment (tboquet) [inline tests/backend/test_sklearn_backend.py]: Codacy Issue found: [Unused variable 'model_deserialized'](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686525116/issues/source?bid=3232749&fileBranchId=3332364#l61)

File: tests/backend/test_sklearn_backend.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
+"""Tests for the sklearn backend"""
+
+
+import numpy as np
+import pytest
```
Comment (tboquet) [inline tests/backend/test_sklearn_backend.py]: Codacy Issue found: [Unused import pytest](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686525116/issues/source?bid=3232749&fileBranchId=3332364#l5)

File: tests/backend/test_sklearn_backend.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+from sklearn.linear_model import Lars
+from sklearn.linear_model import Lasso
+from sklearn.linear_model import LassoLars
+from sklearn.linear_model import LinearRegression
+from sklearn.linear_model import LogisticRegression
+from sklearn.linear_model import OrthogonalMatchingPursuit
+from sklearn.linear_model import Ridge
+
+from alp.appcom.core import Experiment
+from alp.appcom.utils import switch_backend
```
Comment (tboquet) [inline tests/backend/test_sklearn_backend.py]: Codacy Issue found: [Unused switch_backend imported from alp.appcom.utils](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686525116/issues/source?bid=3232749&fileBranchId=3332364#l23)

**Comment by tboquet** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Clones added
============
- tests/backend/test_sklearn_backend.py  44
- src/alp/backend/sklearn_backend.py  2
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/thomas-boquet/python-alp/pullRequest?prid=248711&bid=3332364)

File: tests/backend/test_sklearn_backend.py (hunk 4 above)
Comment (tboquet) [inline tests/backend/test_sklearn_backend.py]: Codacy Issue found: [Unused variable 'model_deserialized'](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686617421/issues/source?bid=3232749&fileBranchId=3332364#l61)

File: tests/backend/test_sklearn_backend.py (hunk 5 above)
Comment (tboquet) [inline tests/backend/test_sklearn_backend.py]: Codacy Issue found: [Unused import pytest](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686617421/issues/source?bid=3232749&fileBranchId=3332364#l5)

File: tests/backend/test_sklearn_backend.py (hunk 6 above)
Comment (tboquet) [inline tests/backend/test_sklearn_backend.py]: Codacy Issue found: [Unused switch_backend imported from alp.appcom.utils](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686617421/issues/source?bid=3232749&fileBranchId=3332364#l23)

File: src/alp/backend/sklearn_backend.py (hunk 3 above)
Comment (tboquet) [inline src/alp/backend/sklearn_backend.py]: Codacy Issue found: [Unused variable 'metric'](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686617414/issues/source?bid=3232749&fileBranchId=3332364#l256)

**Comment by tboquet** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Clones added
============
- tests/backend/test_sklearn_backend.py  44
- src/alp/backend/sklearn_backend.py  2
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/thomas-boquet/python-alp/pullRequest?prid=248711&bid=3332364)

File: src/alp/backend/sklearn_backend.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
+"""
+Adaptor for the sklearn backend
+=============================
+"""
+
+import copy
+import types
```
Comment (tboquet) [inline src/alp/backend/sklearn_backend.py]: Codacy Issue found: [Unused import types](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686816449/issues/source?bid=3232749&fileBranchId=3332364#l7)

File: src/alp/backend/sklearn_backend.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+
+import dill
+import h5py
+import numpy as np
+import six
```
Comment (tboquet) [inline src/alp/backend/sklearn_backend.py]: Codacy Issue found: [Unused import six](https://www.codacy.com/app/thomas-boquet/python-alp/file/2686816449/issues/source?bid=3232749&fileBranchId=3332364#l12)

**Comment by tboquet** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- tests/backend/test_sklearn_backend.py  44
- src/alp/backend/sklearn_backend.py  1
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/thomas-boquet/python-alp/pullRequest?prid=248711&bid=3332364)

**Comment by tboquet** [comment]
Codacy Here is an overview of what got changed by this pull request:

``` diff

Clones added
============
- tests/backend/test_sklearn_backend.py  44
- src/alp/backend/sklearn_backend.py  1

```

See the complete overview on [Codacy](https://www.codacy.com/app/thomas-boquet/python-alp/pullRequest?prid=248711&bid=3332364)

**Comment by tboquet** [comment]
Reviewed 1 of 5 files at r3, 1 of 2 files at r4, 1 of 2 files at r5, 1 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

_[src/alp/appcom/utils.py, line 55 [r7]](https://reviewable.io:443/reviews/tboquet/python-alp/8#-KJLqabd2d10yHHPp9PM:-KJLqabeWLC1ZumcAXhF:1371956750) ([raw file](https://github.com/tboquet/python-alp/blob/7fb8e23f391d51637c34692f5ba7a82c3c6ddc4d/src/alp/appcom/utils.py#L55)):_

> ``` Python
>         if hasattr(backend_m, '__version__'):
>             backend_version = backend_m.__version__
>     elif backend == 'sklearn':
> ```

We should change this: duplicated code.

---

_[src/alp/backend/sklearn_backend.py, line 210 [r7]](https://reviewable.io:443/reviews/tboquet/python-alp/8#-KJLqmdim86mx4bFATfL:-KJLqmdim86mx4bFATfM:344059969) ([raw file](https://github.com/tboquet/python-alp/blob/7fb8e23f391d51637c34692f5ba7a82c3c6ddc4d/src/alp/backend/sklearn_backend.py#L210)):_

> ``` Python
>         """
> 
>     # TODO : métrique
> ```

Delete todos please.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/8)_

**Comment by tboquet** [comment]
Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[src/alp/appcom/utils.py, line 55 [r7]](https://reviewable.io:443/reviews/tboquet/python-alp/8#-KJLqabd2d10yHHPp9PM:-KJLwJgEDpYW9XT8fLly:676834233) ([raw file](https://github.com/tboquet/python-alp/blob/7fb8e23f391d51637c34692f5ba7a82c3c6ddc4d/src/alp/appcom/utils.py#L55)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> We should change this: duplicated code.
> </details>
> LGTM

---

_[src/alp/backend/sklearn_backend.py, line 210 [r7]](https://reviewable.io:443/reviews/tboquet/python-alp/8#-KJLqmdim86mx4bFATfL:-KJLwQ3kJH_ELfQQ_2qq:676834233) ([raw file](https://github.com/tboquet/python-alp/blob/7fb8e23f391d51637c34692f5ba7a82c3c6ddc4d/src/alp/backend/sklearn_backend.py#L210)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> Delete todos please.
> </details>
> LGTM

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/8)_

**Comment by tboquet** [comment]
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/8)_
